File: bar-race-studio/backend/app/api/projects.py
from fastapi import APIRouter, HTTPException

from app.core import db
from app.models.config import RaceConfig
from app.models.project import ProjectDetail, ProjectSummary, SaveProjectRequest

router = APIRouter(prefix="/api/projects", tags=["projects"])


@router.post("")
def save_project(req: SaveProjectRequest):
    project_id = db.save_project(req.name, req.config.dataset_id, req.config.model_dump(mode="json"), req.project_id)
    return {"project_id": project_id}


@router.get("", response_model=list[ProjectSummary])
def list_projects():
    return db.list_projects()


@router.get("/{project_id}", response_model=ProjectDetail)
def get_project(project_id: str):
    project = db.get_project(project_id)
    if project is None:
        raise HTTPException(404, "Project not found")
    project["config"] = RaceConfig(**project["config"])
    return project


@router.delete("/{project_id}")
def delete_project(project_id: str):
    if not db.delete_project(project_id):
        raise HTTPException(404, "Project not found")
    return {"deleted": True}

File: bar-race-studio/frontend/src/api/projects.ts
import { apiRequest } from './client'
import type { RaceConfig } from '../models/RaceConfig'

export interface ProjectSummary {
  id: string
  name: string
  dataset_id: string
  created_at: string
  updated_at: string
}

export interface ProjectDetail extends ProjectSummary {
  config: RaceConfig
}

export function saveProject(name: string, config: RaceConfig, projectId?: string): Promise<{ project_id: string }> {
  return apiRequest('/api/projects', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ name, config, project_id: projectId ?? null }),
  })
}

export function listProjects(): Promise<ProjectSummary[]> {
  return apiRequest('/api/projects')
}

export function getProject(projectId: string): Promise<ProjectDetail> {
  return apiRequest(`/api/projects/${projectId}`)
}
